README
Release 2-2-3 adds the option to delimit LaTeX equations 
by \(...\) instead of $...$ and \[...\] instead of $$...$$

Release 2-2-2 adds conversion of Word's EQ fields
and fixes the missing -lm flag in the gcc compiler call in Makefile

This is rtf2latex2e version 2.2 that

 1. detects text style: bold, italic, color, big, small,...
 2. reads embedded figures: PICT, EMF, GIF, TIFF, WMF, PNG, JPEG
 3. reads tables: simple to semi-complex
 4. equations: converts embedded MathType equations and EQ fields
 5. symbols: converts most greek and math symbols
 6. reads footnotes (not in tables yet)
 7. translates hyperlinks using the hyperref package
 8. converted latex file uses utf8
 9. supports codepages 1250, 1251, 1252, 1254, mac, NeXt
10. converts RTFD (rtf directories)
11. extended preferences
12. internal document page references
13. save all converted files in a separate directory

WHAT YOU WILL GET:
=============
rtf2latex2e converts an RTF file to LaTeX and saves any image
files found inside.  Using pdflatex on the resulting latex
file will require that these images be converted by you to 
some image format supported by pdflatex. 

If you expect a WYSIWYG reproduction of your RTF file, you may be 
disappointed. My main concerns have been translating the essential 
features of the RTF file such as characters, figures, tables, and 
equations (as pictures).  I have largely ignored visual formatting 
such as ruler positions, tabs, paragraph indentations, and other fluff.  
Expect the output LaTeX file to require manual editing to put the finishing 
touches.  I just want to make that task a little easier.  In my opinion, 
expecting a WYSIWYG reproduction is not practical and misses the point 
entirely.

The user has some control over how good the visual fidelity is.  For
example

prompt> rtf2latex2e -t 0 -p 0 test/test.rtf

will produce a file 'test/test.tex' that has no paragraph formatting
or text formatting.  Two other useful variants are

prompt> rtf2latex2e -n test/test.rtf

which makes a file with reasonable latex mark-up.  Fussy details about
paragraph indenting and spacing are left to latex and some work will be
needed to identify paragraphs that should not be indented, etc.,  Finally

prompt> rtf2latex2e -b test/test.rtf

will try all sorts of things to get the formatting correct.  This comes
at the expense of extensive formatting ``noise.'' 

Typeset the rtf2latexDoc.tex for more information, or print out the 
rtf2latexDoc.pdf file. They are in the documentation directory.

The file Release-notes.txt usually has more up-to-date info.

INSTALLATION AND USE:
=====================
rtf2latex2e was developed on under Mac OS X but should compile
under any Unix variant.  Windows users should compile under something
like MinGW.

As of version 2.2.0, rtf2latex2e uses unoconv to convert EMF, WMF, 
and PICT images to PDF files.  unoconv in turn use LibreOffice to
do the actual conversion.  

To build rtf2latex2e, change to the Unix directory directory, type:

make
make test
make install (optional, as root)

By default the executable is installed as /usr/local/bin/rtf2latex2e.
Support files will installed in /usr/local/share/rtf2latex2e

You can change the directory into which rtf2latex2e is installed by editing the 
variable PREFIX in the Makefile. You may need to become 
super-user to install into that directory. If you do not have super-user 
privileges, you can change the PREFIX to somewhere in your home directory, 
say $(HOME)/rtf2latex2e. 

WINDOWS:
--------
Windows users get a pre-compiled binary rtf2latex2e.exe to be run 
from the MS-DOS prompt. 


THE r2l-pref PREFERENCE FILE:
=============================
rtf2latex2e reads a preference file r2l-pref that allows one to specify 
many settings.  Read the comments in the preference file for details.

THE r2l-head file:
==================
Allows you to insert text directly into the preamble of the converted
doument.  

THE r2l-map file:
==================
Allows some direct mapping of paragraph styles to latex commands.  There
are only three styles defined at the moment.  These convert the style
'heading 1' to \section

LEGAL STUFF:
============
(c) Ujwal S. Sathyam and Scott A. Prahl, 1999, 
(c) Scott A. Prahl, 2012

The author provides this software as is and does not take any responsibility 
for the consequences of running it (unless it is good, of course...). This is 
beta software, so treat it as such.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

The writer and driver parts of the code are copyright Ujwal Sathyam.
The reader part of the code was written by Paul DuBois of the Univ. of 
Wisconsin Primate Center. 


doc/Changelog
r288 | wh | 2016-09-17
add missing newline after display equation end

r287 | wh | 2016-09-16
add option to delimit LaTeX equations by \(...\) and \[...\]

r286 | prahl | 2013-04-09
some fixes

r285 | prahl | 2012-10-30
adding .gitignore

r284 | prahl | 2012-10-22
minor tweaks to EQ fields

r283 | prahl | 2012-10-22
fix paragraph bug

r282 | prahl | 2012-10-22
more or less working EQ conversion

r281 | prahl | 2012-10-22
add RTFGetNonWhiteSpaceToken()()

r280 | prahl | 2012-10-16
support EQ braces

r279 | prahl | 2012-10-15
support for EQ fractions

r278 | prahl | 2012-10-15
support for sub and superscripts

r277 | prahl | 2012-10-15
first step towards supporting Word EQ commands

r276 | prahl | 2012-10-07
redefine roundf

r275 | prahl | 2012-08-16
fix parsing of unicode token followed by \'00

r274 | prahl | 2012-08-16
build fix for linux, inlucd math library when linking

r273 | whennings | 2012-07-05
updated emf2pdf.bat to allow blanks in folder names

r272 | whennings | 2012-06-29
updated documentation, Makefile, writer.c for the Windows version 2-2-1

r271 | prahl | 2012-06-29
remove russian-short.tex from makefile

r270 | prahl | 2012-06-29
final ChangeLog update before 2.2.1

r269 | prahl | 2012-06-28
patch future tag to fail better

r268 | prahl | 2012-06-28
bump version number to 2.2.1

r267 | prahl | 2012-06-28
update ChangeLog

r266 | prahl | 2012-06-28
documentation updates

r265 | prahl | 2012-06-28
make sure mandir is present

r264 | prahl | 2012-06-27
correctly identify presence of unoconv

r263 | prahl | 2012-06-27
add mising files to distribution

r262 | prahl | 2012-06-27
use unoconv to convert figures

r261 | prahl | 2012-06-27
change 'convert2pdf' to 'vectorimage2pdf'

r260 | whennings | 2012-06-26
Sorry last version was Makefile for Windows
now uploaded Makefile for Unix

r259 | whennings | 2012-06-26
Makefile modification for MS Windows
added : emf2pdf.bat

r258 | whennings | 2012-06-20
updated Makefile

r257 | whennings | 2012-06-20
writer.c : new wmf/emf to eps converter under MS Windows
added : Readme_Windows.txt

r256 | prahl | 2012-06-17
fix compiler warnings

r255 | prahl | 2012-06-17
support unicode thinspace

r254 | prahl | 2012-06-17
fix bug when footnote is first in doc

r253 | prahl | 2012-06-17
add amssymb package to preamble

r252 | prahl | 2012-06-16
fix fonttable, fix v3 eqn

r251 | prahl | 2012-06-16
Copy permissions during install

r250 | prahl | 2012-06-16
add some more rtf control words

r249 | prahl | 2012-06-16
script to convert files

r248 | prahl | 2012-06-16
add item to list

r247 | prahl | 2012-06-16
better image conversion

r246 | whennings | 2012-03-20
WH: add CC:=gcc to Makefile

r245 | prahl | 2012-03-19
silence useless warning

r244 | prahl | 2012-03-19
remove unused files and fix image insertion

r243 | prahl | 2012-03-19
don't save unused images

r242 | prahl | 2012-03-19
fix pict export when header is present

r241 | prahl | 2012-03-17
tabs->spaces

r240 | prahl | 2012-03-17
make ole equations more robust

r239 | prahl | 2012-03-17
add web files to dist

r238 | prahl | 2012-03-14
tweak title

r237 | prahl | 2012-03-14
html errors and style update

r236 | prahl | 2012-03-13
add rsync install to Makefile

r235 | prahl | 2012-03-13
add MTEF html files

r234 | prahl | 2012-03-13
add hand-crafted web files

r233 | prahl | 2012-03-11
add man file, remove SWP file

r232 | prahl | 2012-03-11
tweak help message

r231 | prahl | 2012-03-11
more tweaks to Makefile

r230 | prahl | 2012-03-11
adhere to www.gnu.org/prep/standards/

r229 | prahl | 2012-03-11
fix prototype issues in rtfprep

r228 | prahl | 2012-03-11
fix remaining prototype issues

r227 | prahl | 2012-03-11
remove redundant mkdir definition in main.c

r226 | prahl | 2012-03-11
fix prototypes and shadows in eqn.c

r225 | prahl | 2012-03-11
fix prototypes in rtf.h

r224 | prahl | 2012-03-11
update most prototypes in tokenscan

r223 | prahl | 2012-03-11
support blacksquare and document process

r222 | prahl | 2012-03-11
account for [xx] after \linebreak

r221 | prahl | 2012-03-11
handle CR before picture data

r220 | prahl | 2012-03-10
quiet warnings in table code

r219 | prahl | 2012-03-10
handle ole header differently

r218 | prahl | 2012-03-06
support for \uc0

r217 | prahl | 2012-02-12
testing git svn

r216 | prahl | 2011-11-26
start of multirow support

r215 | prahl | 2011-11-25
trivial changes to try and understand the vertically
merged cell problem

r214 | prahl | 2011-11-23
rework tables so that column alignment works better
now table.tex works up to table 6.

r213 | prahl | 2011-11-17
remove ancient comments that really don't help

r212 | prahl | 2011-11-17
when emitting latex without preamble, still include
necessary color definitions.
Add a sentence to the r2l-pref file to emphasize that
PICT conversion no longer works under Lion.

r211 | prahl | 2011-11-17
fixes so that fonttable and colortable ignore some
new keywords/tokens found in Word 2008 for Mac RTF

r210 | prahl | 2011-11-16
minor changes to try and get new rtf file to convert

r209 | prahl | 2011-11-09
add -f switch to avoid including latex preamble

r208 | whennings | 2011-09-13
WH: Makefile adaptation for Windows

r207 | prahl | 2011-09-06
Clean up Makefile as was done for latex2rtf.
Eliminate unused variables
Remove binary before installing on top

r206 | prahl | 2011-08-26
Change PREFIX to DESTDIR to follow GNU standard
http://www.gnu.org/prep/standards/html_node/DESTDIR.html

r205 | prahl | 2011-05-10
properly escape ampersand in equations.  Use \mathop instead
of \limfun for some operators e.g., var()

r204 | prahl | 2011-03-06
fix NextGraphics path

r203 | prahl | 2011-03-06
when translating .rtfd files, always create a new directory
and copy all the figures to the new directory.

r202 | prahl | 2011-03-05
scale column width for oversize tables

r201 | prahl | 2011-02-28
handle text in tables very similarly to text handled elsewhere
set default conversion of tables to avoid a lot of extra noise

src/main.c
/*
 * RTF-to-LaTeX2e translation driver code.
 * (c) 1999 Ujwal S. Sathyam
 * (c) 2011 Scott Prahl
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

#include    <stdio.h>
#include    <string.h>
#include    <stdlib.h>
#include    <ctype.h>
#include    <stdarg.h>
#include    <stdint.h>
#include    <sys/stat.h>

#include     "rtf.h"
#include     "mygetopt.h"
#include     "rtf2latex2e.h"
#include     "eqn.h"
#include     "init.h"
boolean ConvertRawEquationFile(char *rawFileName); /* from writer.c */

FILE         *ifp, *ofp;

char  *g_library_path        = NULL;
int   g_little_endian        = 0;
int   g_debug_level          = 0;
int   g_object_width         = 0;
int   g_create_new_directory = 0;

int   g_eqn_insert_image     = 0;
int   g_eqn_keep_file        = 0;
int   g_eqn_insert_name      = 0;
int   g_shouldIncludePreamble= 1;

enum INPUT_FILE_TYPE g_input_file_type;

/* Figure out endianness of machine.  Needed for OLE & graphics support */
static void 
SetEndianness(void)
{
    unsigned int    endian_test = (unsigned int) 0xaabbccdd;
    unsigned char   endian_test_char = *(unsigned char *) &endian_test;
    if (endian_test_char == 0xdd)
        g_little_endian = 1;
}


static void 
print_version(void)
{
    fprintf(stdout, "rtf2latex2e %s\n\n", VERSION);
    fprintf(stdout, "Copyright (C) 2012 Free Software Foundation, Inc.\n");
    fprintf(stdout, "This is free software; see the source for copying conditions.  There is NO\n");
    fprintf(stdout, "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\n");
}

static void 
print_usage(void)
{
    fprintf(stdout, "rtf2latex2e %s - convert RTF to LaTeX.\n", VERSION);
    fprintf(stdout, "Options:\n");
    fprintf(stdout, "  -h               display help\n");
    fprintf(stdout, "  -b               best attempt at matching RTF formatting\n");
    fprintf(stdout, "  -f               fractional document, omit latex preamble\n");
    fprintf(stdout, "  -D               make a new directory for latex and extracted images\n");
    fprintf(stdout, "  -e #             equation conversion options\n");
    fprintf(stdout, "      -e 1              convert to latex\n");
    fprintf(stdout, "      -e 2              insert image\n");
    fprintf(stdout, "      -e 4              keep intermediate eqn file\n");
    fprintf(stdout, "      -e 8              insert eqn file name in latex document\n");
    fprintf(stdout, "      -e 16             delimit eqns by \\(...\\) and \\[...\\]\n");
    fprintf(stdout, "  -n               natural latex formatting ... easiest to edit\n");
    fprintf(stdout, "  -p #             paragraph conversion options\n");
    fprintf(stdout, "      -p 1              'heading 1' style -> '\\section{}'\n");
    fprintf(stdout, "      -p 2              convert indenting\n");
    fprintf(stdout, "      -p 4              convert space between paragraphs\n");
    fprintf(stdout, "      -p 8              convert line spacing\n");
    fprintf(stdout, "      -p 16             convert margins\n");
    fprintf(stdout, "      -p 32             convert alignment\n");
    fprintf(stdout, "  -P path          path to preferred preference directory\n");
    fprintf(stdout, "  -t #             text conversion options\n");
    fprintf(stdout, "      -t 1              convert text size\n");
    fprintf(stdout, "      -t 2              convert text color\n");
    fprintf(stdout, "      -t 4              convert text formatting\n");
    fprintf(stdout, "      -t 8              replace tabs with spaces\n");
    fprintf(stdout, "  -T #             table conversion options\n");
    fprintf(stdout, "      -T 1              keep column widths\n");
    fprintf(stdout, "      -T 2              keep column alignment\n");
    fprintf(stdout, "  -v               version information\n");
    fprintf(stdout, "\n");
    fprintf(stdout, "Examples:\n");
    fprintf(stdout, "  rtf2latex2e foo              convert foo.rtf to foo.tex\n");
    fprintf(stdout, "  rtf2latex2e -n foo           minimal mark-up\n");
    fprintf(stdout, "  rtf2latex2e -e 15 foo        help identify failed equation conversion\n");
    fprintf(stdout, "  rtf2latex2e foo-eqn003.eqn   debug third equation (after above command)\n");
    fprintf(stdout, "  rtf2latex2e -D foo           put foo.tex and images in foo-latex dir\n");
    fprintf(stdout, "  rtf2latex2e foo.rtfd         convert to foo.rtfd/TXT.tex\n");
    fprintf(stdout, "\n");
    fprintf(stdout, "Preference Directory:\n");
    fprintf(stdout, "  Default  = '%s'\n", (PREFS_DIR) ? PREFS_DIR : "not defined");
    fprintf(stdout, "  $RTFPATH = '%s'\n", (getenv("RTFPATH")) ? getenv("RTFPATH") : "not defined");
    fprintf(stdout, "\n");
    fprintf(stdout, "Report bugs at http://sourceforge.net/projects/rtf2latex2e/\n\n");
    exit(1);
}

static enum INPUT_FILE_TYPE identify_filename(char *name)
{
    int len;
    
    if (!name) return TYPE_UNKNOWN;
    
    len = strlen(name);
    
    if (len > 5 && strcasecmp(name+len-5,".rtfd")==0)
        return TYPE_RTFD;
        
    if (len > 4 && strcasecmp(name+len-4,".rtf")==0)
        return TYPE_RTF;
        
    if (len > 4 && strcasecmp(name+len-4,".eqn")==0)
        return TYPE_EQN;
        
    if (len > 4 && strcasecmp(name+len-4,".rqn")==0)
        return TYPE_RAWEQN;

    return TYPE_UNKNOWN;
}


static char * establish_filename(char * name)
{
    FILE *fp;
    char *s;
    int len;
    
    g_input_file_type = TYPE_UNKNOWN;
    if (!name) return NULL;
    
    /* try .rtfd case first because simple fopen() test below misses this case */
    len = strlen(name) - 5;
    if (len > 0 && strcasecmp(name + len,".rtfd") == 0) {
        s = append_file_to_path(name, "TXT.rtf");
        fp = fopen(name, "rb");
        if (fp) {
            fclose(fp);
            g_input_file_type = TYPE_RTFD;
            return s;
        }
        free(s);
        fprintf(stderr,"* RTFD directory found, but no TXT.rtf file inside!\n");
        return NULL;
    }
    
    fp = fopen(name, "rb");
    if (fp) {
        fclose(fp);
        g_input_file_type = identify_filename(name);
        return strdup(name);
    }
    
    s = strdup_together(name, ".rtf");
    fp = fopen(s, "rb");
    if (fp) {
        fclose(fp);
        g_input_file_type = TYPE_RTF;
        return s;
    }

    free(s);
    s = strdup_together(name, ".rtfd");
    fp = fopen(s, "rb");
    if (fp) {
        fclose(fp);
        free(s);
        s = strdup_together(name, ".rtfd/TXT.rtf");
        fp = fopen(s, "rb");
        if (fp) {
            fclose(fp);
            g_input_file_type = TYPE_RTFD;
            return s;
        }
        
        return s;
        fprintf(stderr,"* RTFD directory found, but no TXT.rtf file inside!\n");
        return NULL;
    }

    free(s);
    s = strdup_together(name, ".eqn");
    fp = fopen(s, "rb");
    if (fp) {
        fclose(fp);
        g_input_file_type = TYPE_EQN;
        return s;
    }

    free(s);
    return NULL;
}


static char * short_name(char *path)
{
    char *s;
    s = strrchr(path,PATH_SEP);
    if (s == NULL) 
        return strdup(path);
    else
        return strdup(s+1);
}

/* initial name for translated file will end in .ltx.  Later
   the file will be rewritten to a name that ends with .tex.
   the windows rename() function is horrible and so we do this
   dance instead (see EndLatexFile in writer.c)
 */
static char * make_output_filename(char * name)
{
    char *s, *dir, *file, *out;
    if (!name) return NULL;
    
    /* always create a new folder for RTFD files */
    if (g_input_file_type == TYPE_RTFD) {

        dir = malloc(strlen(name)+strlen("-latex"));
        strcpy(dir,name);
        s = strstr(dir, ".rtfd");
        if (!s) 
        	s = strstr(dir, ".RTFD");

        if (s)
        	strcpy(s,"-latex");
        else
        	strcat(dir,"-latex");
        
        mkdir(dir, 0755);
        out = append_file_to_path(dir,"TXT.ltx");
        return out;
    }

    if (!g_create_new_directory || g_input_file_type == TYPE_EQN || g_input_file_type == TYPE_RAWEQN) {
        s = strdup(name);
        strcpy(s+strlen(s)-4, ".ltx");
        return s;
    } 
    
    if (g_input_file_type == TYPE_RTF || g_input_file_type == TYPE_RTFD) {
        file = short_name(name);        
        strcpy(file+strlen(file)-4, ".ltx");

        dir = malloc(strlen(name)+strlen("-latex"));
        strcpy(dir,name);
        strcpy(dir+strlen(dir)-4,"-latex");
        
        mkdir(dir, 0755);
        out = append_file_to_path(dir,file);
        return out;
    }
    
    return NULL;
}

void ExamineToken(void);

int 
main(int argc, char **argv)
{
    char            c, *input_filename, *output_filename;
    int             fileCounter, cli_paragraph, cli_text, cli_equation, cli_table;
    int             g_eqn_delim_brackets;
    long            cursorPos;
    extern char    *optarg;
    extern int      optind;

    SetEndianness();
    cli_paragraph = -1;
    cli_text = -1;
    cli_equation = -1;
    cli_table = -1;

    g_eqn_delim_brackets = 0;
    eqn_start_inline = strdup("$");
    eqn_end_inline = strdup(" $");
    eqn_start_display = strdup(" $$");
    eqn_end_display = strdup("$$\n");

    while ((c = my_getopt(argc, argv, "bDe:fhnp:P:t:T:v")) != EOF) {
        switch (c) {

        case 'b':
            cli_paragraph = 63;
            cli_text = 7;
            cli_equation = 1;
            cli_table = 3;
            break;

        case 'D':
            g_create_new_directory = 1;
            break;

        case 'e':
            sscanf(optarg, "%d", &cli_equation);
            break;

        case 'f':
            g_shouldIncludePreamble = 0;
            break;

        case 'n':
            cli_paragraph = 33;
            cli_text = 12;
            cli_equation = 1;
            cli_table = 0;
            break;

        case 'p':
            sscanf(optarg, "%d", &cli_paragraph);
            break;

        case 'P':   /* -P path/to/pref */
            g_library_path = strdup(optarg);
            break;

        case 't':
            sscanf(optarg, "%d", &cli_text);
            break;

        case 'T':
            sscanf(optarg, "%d", &cli_table);
            break;

        case 'v':
        case 'V':
            print_version();
            return (0);

        case 'h':
        case '?':
        default:
            print_usage();
        }
    }

    argc -= optind;
    argv += optind;

    if (!argc) print_usage();
    InitConverter();

    if (cli_paragraph >= 0) {
        prefs[pConvertParagraphStyle]      = cli_paragraph & 1;
        prefs[pConvertParagraphIndent]     = cli_paragraph & 2;
        prefs[pConvertInterParagraphSpace] = cli_paragraph & 4;
        prefs[pConvertLineSpacing]         = cli_paragraph & 8;
        prefs[pConvertParagraphMargin]     = cli_paragraph & 16;
        prefs[pConvertParagraphAlignment]  = cli_paragraph & 32;
    }

    if (cli_text >= 0) {
        prefs[pConvertTextSize]  = cli_text & 1;
        prefs[pConvertTextColor] = cli_text & 2;
        prefs[pConvertTextForm]  = cli_text & 4;
        prefs[pConvertTextNoTab] = cli_text & 8;
    }

    if (cli_equation >= 0) {
        prefs[pConvertEquation] = cli_equation & 1;
        g_eqn_insert_image      = cli_equation & 2;
        g_eqn_keep_file         = cli_equation & 4;
        g_eqn_insert_name       = cli_equation & 8;
        g_eqn_delim_brackets    = cli_equation & 16;
    }
    if  (g_eqn_delim_brackets) {
        eqn_start_inline = strdup("\\(");
        eqn_end_inline = strdup(" \\)");
        eqn_start_display = strdup("\\[");
        eqn_end_display = strdup(" \\]\n");
        }
    
    if (cli_table >= 0) {
        prefs[pConvertTableWidths]    = cli_table & 1;
        prefs[pConvertTableAlignment] = cli_table & 2;
        prefs[pConvertTableBorders]   = cli_table & 4;
    }

    for (fileCounter = 0; fileCounter < argc; fileCounter++) {

        RTFInit();
        
        input_filename = establish_filename(argv[fileCounter]);
    	
    	if (g_input_file_type == TYPE_RTFD)
    		g_create_new_directory = 1;
    		
        if (g_input_file_type == TYPE_UNKNOWN) {
            if (!input_filename)
                RTFMsg("* Skipping non-existent file '%s'\n", argv[fileCounter]);
            else 
                RTFMsg("* Skipping non-RTF file '%s'\n", argv[fileCounter]);
            if (input_filename) free(input_filename);
            continue;
        }
        
        ifp = fopen(input_filename, "rb");
        if (!ifp) {
            RTFPanic("* Cannot open input file %s\n", input_filename);
            exit(1);
        }
        
        RTFSetInputName(input_filename);
        RTFSetStream(ifp);

        /* look at second token to check if input file is of type rtf */
        if (g_input_file_type != TYPE_EQN && g_input_file_type != TYPE_RAWEQN) {
            cursorPos = ftell(ifp);
            RTFGetToken();
            RTFGetToken();
            if (!RTFCheckCMM(rtfControl, rtfVersion, rtfVersionNum)) {
                RTFMsg("* Yikes! '%s' is not actually a RTF file!  Skipping....\n", input_filename);
                fclose(ifp);
                free(input_filename);
                continue;
            }
            fseek(ifp, cursorPos, 0);
        }
        RTFInit();
        
        output_filename = make_output_filename(input_filename);

        ofp = fopen(output_filename, "wb+");
        if (!ofp) {
            RTFMsg("Cannot open output file %s\n", output_filename);
            free(input_filename);
            free(output_filename);
            continue;
        }
        RTFSetOutputName(output_filename);
        RTFSetOutputStream(ofp);

        fprintf(stderr, "Processing %s\n", input_filename);
        if (BeginLaTeXFile()) {
            if (g_input_file_type == TYPE_EQN) 
                (void) ConvertEquationFile(input_filename);
            else if (g_input_file_type == TYPE_RAWEQN) {
                (void) ConvertRawEquationFile(input_filename);
            } else

                RTFRead();
            EndLaTeXFile(); /* closes ofp */
        }
        
        fclose(ifp);
        free(input_filename);
        free(output_filename);
    }

    return (0);
}
